### PR TITLE
Create a service for dashboard agent

### PR DESCRIPTION
**Pull request by brucez-anyscale**
## Why are these changes needed?
Ray nightly moves serve endpoints to dashboard agent. This pr adopts to it.

We need to send the dashboard agent listen port to ray as start param.
The serve deployment related requests will now send to a new dashboard agent service maintained by RayCluster controller..

Manual test in EKS.
Screen Shot 2022-06-27 at 4 30 43 PM

## Checks

- [X] I've made sure the tests are passing. 
- Testing Strategy
   - [X] Unit tests
   - [X] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/common/pod.go [hunk 1]
```diff
@@ -417,6 +418,16 @@ func concatenateContainerCommand(nodeType rayiov1alpha1.RayNodeType, rayStartPar
 		}
 	}
 
+	if _, ok := rayStartParams["dashboard-agent-listen-port"]; !ok {
+		rayStartParams["dashboard-agent-listen-port"] = fmt.Sprint(DefaultDashboardAgentListenPort)
+		for _, port := range ports {
+			if port.Name == "dashboard-agent" {
+				rayStartParams["dashboard-agent-listen-port"] = fmt.Sprint(port.ContainerPort)
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/common/pod.go:425]: Is dashboard-agent-listen-port a new arg added to Ray nightly?
Is this going to crash earlier versions of Ray?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -198,7 +201,49 @@ func (r *RayClusterReconciler) reconcileServices(instance *rayiov1alpha1.RayClus
...
+		// if the service cannot be created we return the error and requeue
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *RayClusterReconciler) reconcileDashBoardServices(instance *rayiov1alpha1.RayCluster) error {
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/raycluster_controller.go:214]: Would it be desirable to use some common code with the existing headService reconciler?
(It might not be, just asking.)

File: ray-operator/controllers/ray/common/pod.go (hunk 1 above)
Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:425]: If so, one approach would be to add a toggle to turn the new behavior on/off.
If the toggle is on, it could potentially also auto-configure the port, so that the user has to do less.

Comment (brucez-anyscale) [reply ray-operator/controllers/ray/common/pod.go:425]: It already exists

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 2 above)
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/raycluster_controller.go:214]: We could let them share a method with some input parameter.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/raycluster_controller.go:214]: Let's do that if it's not too awkward.

File: ray-operator/controllers/ray/common/pod.go (hunk 1 above)
Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:425]: Ah, yep it's been there since Ray 1.6.0 -- early August 2021.
Should be fine -- I don't think there's need to keep compatibility that far back.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 3]
```diff
@@ -119,6 +119,9 @@ func (r *RayClusterReconciler) Reconcile(ctx context.Context, request ctrl.Reque
 	if err := r.reconcileServices(instance); err != nil {
 		return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
 	}
+	if err := r.reconcileDashBoardServices(instance); err != nil {
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/raycluster_controller.go:122]: nit: reconcielDashboardServices

File: ray-operator/controllers/ray/common/pod.go [hunk 4]
```diff
@@ -417,6 +418,16 @@ func concatenateContainerCommand(nodeType rayiov1alpha1.RayNodeType, rayStartPar
 		}
 	}
 
+	if _, ok := rayStartParams["dashboard-agent-listen-port"]; !ok {
+		rayStartParams["dashboard-agent-listen-port"] = fmt.Sprint(DefaultDashboardAgentListenPort)
+		for _, port := range ports {
+			if port.Name == "dashboard-agent" {
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/common/pod.go:424]: Hmm, the UI is that the functionality is turned on/off by presence of a port named "dashboard-agent".
This feels a little awkward. Should we just auto-configure the port with a default?

**Comment by DmitriGekhtman** [comment]
> Is this only available on Ray master? I notice we have a few master changes needs to sync in KubeRay repo and we might need a better strategy to managed versions and compatibilities.

This PR would break compatibility with Ray 1.5.0 and earlier -- this is probably fine, but we should document it.
(When we discussed compatibility earlier, we aimed to make each KubeRay release test-ably compatible with the last three Ray minor versions.)

The use-case for a service targeting this particular port on the workers only makes sense on master.
(We've just moved serve endpoints to the worker dashboard agents.)

It does feel a bit odd to always create a K8s service for a completely new functionality. Maybe we can add a flag to the Ray CR that the RayService controller would always flip on?

**Comment by brucez-anyscale** [comment]
> > Is this only available on Ray master? I notice we have a few master changes needs to sync in KubeRay repo and we might need a better strategy to managed versions and compatibilities.
> 
> This PR would break compatibility with Ray 1.5.0 and earlier -- I this is fine, but we should document it. (When we discussed compatibility earlier, we aimed to make each KubeRay release test-ably compatible with the last three Ray minor versions.)
> 
> The use-case for a service targeting this particular port on the workers only makes sense on master. (We've just moved serve endpoints to the worker dashboard agents.)
> 
> It does feel a bit odd to always create a K8s service for a completely new functionality. Maybe we can add a flag to the Ray CR that the RayService controller would always flip on?

Let me think about how to make it backward compatible. For the latest Ray, it is ok we do not create this new service, the head service starting with the new target port can still work but without replicas.

File: ray-operator/controllers/ray/common/pod.go (hunk 4 above)
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/common/pod.go:424]: For this pr, let's assume it defined in RayService yaml. I can improve it in the follow up pr.

**Comment by DmitriGekhtman** [comment]
This PR has been changed so that the ray start parameter `dashboard-agent-listen-port` is not auto-filled, which if I understand right, makes the changes in this PR fully backwards compatible.

@brucez-anyscale Could you explain again what happens if the ray start parameter `dashboard-agent-listen-port` is not set in the config?

Questions for other reviewers cc @Jeffwan : is it ok that the RayCluster reconciler now always creates a new service for the dashboard agent? I guess even if the user is not using Ray serve, it's harmless to keep the extra dashboard agent service around.

File: ray-operator/config/manager/kustomization.yaml [hunk 5]
```diff
@@ -5,5 +5,6 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
 commonLabels:
-  app.kubernetes.io/name: kuberay
   app.kubernetes.io/component: kuberay-operator
+  app.kubernetes.io/name: kuberay
```
Comment (DmitriGekhtman) [inline ray-operator/config/manager/kustomization.yaml:9]: Any reason for this diff, or is it accidental?

File: ray-operator/config/rbac/role.yaml [hunk 6]
```diff
@@ -3,6 +3,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
+  creationTimestamp: null
```
Comment (DmitriGekhtman) [inline ray-operator/config/rbac/role.yaml:6]: Why is this needed?

File: ray-operator/config/samples/ray_v1alpha1_rayservice.yaml [hunk 7]
```diff
@@ -36,7 +36,7 @@ spec:
         rayActorOptions:
           numCpus: 0.1
   rayClusterConfig:
-    rayVersion: '1.12.1' # should match the Ray version in the image of the containers
+    rayVersion: 'nightly' # should match the Ray version in the image of the containers
```
Comment (DmitriGekhtman) [inline ray-operator/config/samples/ray_v1alpha1_rayservice.yaml:39]: I think we've established this, but we'll want to remove "nightly" after the 2.0.0 release or even once we have a release candidate.

File: ray-operator/config/rbac/role.yaml (hunk 6 above)
Comment (brucez-anyscale) [reply ray-operator/config/rbac/role.yaml:6]: This is auto generated. Not sure the reason

File: ray-operator/config/samples/ray_v1alpha1_rayservice.yaml (hunk 7 above)
Comment (brucez-anyscale) [reply ray-operator/config/samples/ray_v1alpha1_rayservice.yaml:39]: right. Now we only works with nightly. Later we can change it to 2.0.0

File: ray-operator/config/rbac/role.yaml (hunk 6 above)
Comment (DmitriGekhtman) [reply ray-operator/config/rbac/role.yaml:6]: This doesn't do anything, so I guess it's fine.

File: ray-operator/controllers/ray/rayservice_controller.go [hunk 8]
```diff
@@ -324,7 +323,7 @@ func (r *RayServiceReconciler) shouldPrepareNewRayCluster(rayServiceInstance *ra
 		// Prepare new RayCluster if:
 		// 1. No active and pending cluster
 		// 2. No pending cluster, and the active RayCluster has changed.
-		if activeRayCluster == nil || !reflect.DeepEqual(activeRayCluster.Spec, rayServiceInstance.Spec.RayClusterSpec) {
+		if activeRayCluster == nil || !utils.CompareJsonStruct(activeRayCluster.Spec, rayServiceInstance.Spec.RayClusterSpec) {
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/rayservice_controller.go:326]: Could you explain this diff?
I guess you could have different objects that serialize to the same JSON and this prevents you from considering them different.
Should we always be using this method in place of DeepEqual?

File: ray-operator/controllers/ray/common/constant.go [hunk 9]
```diff
@@ -2,12 +2,16 @@ package common
...
-	RayIDLabelKey        = "ray.io/identifier"
+	RayServiceLabelKey                 = "ray.io/service"
+	RayClusterLabelKey                 = "ray.io/cluster"
+	RayNodeTypeLabelKey                = "ray.io/node-type"
+	RayNodeGroupLabelKey               = "ray.io/group"
+	RayNodeLabelKey                    = "ray.io/is-ray-node"
+	RayIDLabelKey                      = "ray.io/identifier"
+	RayClusterDashboardServiceLabelKey = "ray.io/cluster-dashboard"
+
+	EnableAgentServiceKey   = "enableAgentService"
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/common/constant.go:13]: How about "ray.io/enableAgentService"?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 10]
```diff
@@ -125,7 +126,13 @@ func (r *RayClusterReconciler) Reconcile(ctx context.Context, request ctrl.Reque
...
 		return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
 	}
-	if err := r.reconcileServices(instance); err != nil {
+	if err := r.reconcileServices(instance, common.HeadService); err != nil {
+		if updateErr := r.updateClusterState(instance, rayiov1alpha1.Failed); updateErr != nil {
+			log.Error(updateErr, "RayCluster update state error", "cluster name", request.Name)
+		}
+		return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
+	}
+	if err := r.reconcileServices(instance, common.AgentService); err != nil {
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/raycluster_controller.go:135]: Should we guard this behind the flag in the annotations too?
i.e. do not create the agent service if the flag is not set to "true"?

Comment (brucez-anyscale) [reply ray-operator/controllers/ray/raycluster_controller.go:135]: sounds good

File: ray-operator/controllers/ray/rayservice_controller.go (hunk 8 above)
Comment (brucez-anyscale) [reply ray-operator/controllers/ray/rayservice_controller.go:326]: Suggest to use this for json struct comparison. The deepEqual will fail in some cases.

**Review by DmitriGekhtman — APPROVED** [review]
Thanks!
